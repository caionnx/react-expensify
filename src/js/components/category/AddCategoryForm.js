import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import slugify from 'slugify'
import { startAddCategory } from '../../actions/categories'

export class AddCategoryForm extends React.Component {
  state = {
    error: '',
    newCategory: ''
  }

  onAddCategory = (event) => {
    event.preventDefault()
    const value = this.state.newCategory
    const id = slugify(value, { replacement: '-', remove: null, lower: true })

    if (id === '') {
      this.setState(() => ({ error: 'Please provide category name.' }))
      return
    }
    if (this.props.categories.filter((item) => item.id === id).length) {
      this.setState(() => ({ error: 'Category already registered.' }))
      return
    }

    this.setState(() => ({ error: '', newCategory: '' }))
    this.props.startAddCategory({ id, value })
  }

  onInputChange = (event) => {
    this.setState({ newCategory: event.target.value })
  }

  render () {
    const { error, newCategory } = this.state

    return (
      <form onSubmit={this.onAddCategory}>
        {error && <p className='form__error'>{error}</p>}
        <div className='input-group'>
          <div className='input-group__item input-group__item--expanded'>
            <input
              value={newCategory}
              onChange={this.onInputChange}
              className='text-input'
              placeholder='New Category'
              type='text' />
          </div>
          <div className='input-group__item'>
            <button className='button button--expanded'>Add Category</button>
          </div>
        </div>
      </form>
    )
  }
}

AddCategoryForm.propTypes = {
  categories: PropTypes.array,
  startAddCategory: PropTypes.func.isRequired
}

const mapStateToProps = ({ categories }) => ({
  categories
})

const mapDispatchToProps = (dispatch) => ({
  startAddCategory: (category) => dispatch(startAddCategory(category))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddCategoryForm)
