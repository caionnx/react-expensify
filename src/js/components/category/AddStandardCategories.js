import React from 'react'
import PropTypes from 'prop-types'

export const AddStandardCategories = ({ buttonAction }) => (
  <div className='list-item list-item--message'>
    <div>
      <h3 className='list-item__title'>You dont have any category registered.</h3>
      <button className='button' onClick={() => buttonAction()}>
        Start with standard categories
      </button>
    </div>
  </div>
)

AddStandardCategories.propTypes = {
  buttonAction: PropTypes.func.isRequired
}

export default AddStandardCategories
