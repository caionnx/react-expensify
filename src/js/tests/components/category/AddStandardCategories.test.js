import React from 'react'
import { shallow } from 'enzyme'
import { AddStandardCategories } from '../../../components/category/AddStandardCategories'

let wrapper, buttonAction

beforeEach(() => {
  buttonAction = jest.fn()
  wrapper = shallow(<AddStandardCategories buttonAction={buttonAction} />)
})

test('should render AddStandardCategories correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should run buttonAction prop function', () => {
  const button = wrapper.find('button')
  button.simulate('click')

  expect(buttonAction).toHaveBeenCalled()
})
