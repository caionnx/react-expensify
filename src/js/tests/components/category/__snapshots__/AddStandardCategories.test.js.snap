// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddStandardCategories correctly 1`] = `
<div
  className="list-item list-item--message"
>
  <div>
    <h3
      className="list-item__title"
    >
      You dont have any category registered.
    </h3>
    <button
      className="button"
      onClick={[Function]}
    >
      Start with standard categories
    </button>
  </div>
</div>
`;
